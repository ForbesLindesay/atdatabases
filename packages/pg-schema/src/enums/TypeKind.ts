enum TypeKind {
  // N.B. Array is not really a type in postgres,
  // instead Array is just a special kind of base
  // type
  Array = 'array',
  Base = 'b',
  // (e.g., a table's row type)
  Composite = 'c',
  // domains alias other types, with optional constraints
  Domain = 'd',
  Enum = 'e',
  Pseudo = 'p',
}

export default TypeKind;
