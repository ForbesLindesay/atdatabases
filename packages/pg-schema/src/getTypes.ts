import {Connection, sql} from '@databases/pg';
import TypeCateogry from './enums/TypeCategory';
import TypeKind from './enums/TypeKind';
import getAttributes, {Attribute} from './getAttributes';
import getEnumValues from './getEnumValues';

export interface TypeQuery {
  schemaID?: number;
  schemaName?: string;
  typeID?: number;
  typeName?: string;
  category?: TypeCateogry;
}
export interface TypeRecord {
  schemaID: number;
  schemaName: string;
  typeID: number;
  typeName: string;
  kind: TypeKind;
  category: TypeCateogry;
  /**
   * classID if composite type
   */
  classID?: number;
  subtypeID?: number;
  subtypeName?: string;
  basetypeID?: number;
  basetypeName?: string;
  comment: string | null;
}

export interface TypeBase {
  schemaID: number;
  schemaName: string;
  typeID: number;
  typeName: string;
  kind: TypeKind;
  category: TypeCateogry;
  comment: string | null;
}

export interface ArrayType extends TypeBase {
  kind: TypeKind.Array;
  subtypeID: number;
  subtypeName: string;
}
export interface BaseType extends TypeBase {
  kind: TypeKind.Base;
  subtypeID?: number;
  subtypeName?: string;
}
export interface CompositeType extends TypeBase {
  kind: TypeKind.Composite;
  classID: number;
  attributes: Attribute[];
}
export interface DomainType extends TypeBase {
  kind: TypeKind.Domain;
  basetypeID: number;
  basetypeName: string;
}
export interface EnumType extends TypeBase {
  kind: TypeKind.Enum;
  values: string[];
}
export interface PseudoType extends TypeBase {
  kind: TypeKind.Pseudo;
}

export type Type =
  | ArrayType
  | BaseType
  | CompositeType
  | DomainType
  | EnumType
  | PseudoType;

export default async function getTypes(
  connection: Connection,
  query: TypeQuery = {},
): Promise<Type[]> {
  const conditions = typeQuery(query);

  const typeRecords = (await connection.query(sql`
    SELECT
      ns.oid AS "schemaID",
      ns.nspname AS "schemaName",
      ty.oid as "typeID",
      ty.typname AS "typeName",
      ty.typtype AS "kind",
      ty.typcategory AS "category",
      ty.typrelid AS "classID",
      subt.oid as "subtypeID",
      subt.typname AS "subtypeName",
      baset.oid as "basetypeID",
      baset.typname AS "basetypeName",
      
      obj_description(ty.oid, 'pg_type') as "comment"
    FROM pg_catalog.pg_type ty
    INNER JOIN pg_catalog.pg_namespace ns
      ON (ty.typnamespace = ns.oid)
    LEFT OUTER JOIN pg_catalog.pg_type subt
      ON (ty.typelem = subt.oid)
    LEFT OUTER JOIN pg_catalog.pg_type baset
      ON (ty.typbasetype = baset.oid)
    ${conditions.length ? sql`WHERE ${sql.join(conditions, ' AND ')}` : sql``}
  `)) as TypeRecord[];

  return Promise.all(
    typeRecords.map(
      async (tr): Promise<Type> => {
        const base: TypeBase = {
          schemaID: tr.schemaID,
          schemaName: tr.schemaName,
          typeID: tr.typeID,
          typeName: tr.typeName,
          kind: tr.kind,
          category: tr.category,
          comment: tr.comment,
        };
        switch (base.kind) {
          case TypeKind.Array:
          case TypeKind.Base:
            if (tr.category === TypeCateogry.Array) {
              return {
                ...base,
                kind: TypeKind.Array,
                subtypeID: tr.subtypeID!,
                subtypeName: tr.subtypeName!,
              };
            } else {
              return {
                ...base,
                kind: TypeKind.Base,
                subtypeID: tr.subtypeID,
                subtypeName: tr.subtypeName,
              };
            }
          case TypeKind.Composite:
            return {
              ...base,
              kind: TypeKind.Composite,
              classID: tr.classID!,
              attributes: await getAttributes(connection, {
                classID: tr.classID!,
              }),
            };
          case TypeKind.Domain:
            return {
              ...base,
              kind: TypeKind.Domain,
              basetypeID: tr.basetypeID!,
              basetypeName: tr.basetypeName!,
            };
          case TypeKind.Enum:
            return {
              ...base,
              kind: TypeKind.Enum,
              values: (await getEnumValues(connection, {
                typeID: tr.typeID,
              })).map(v => v.value),
            };
          case TypeKind.Pseudo:
            return {
              ...base,
              kind: TypeKind.Pseudo,
            };
          default:
            const kind: never = base.kind;
            return {
              ...base,
              kind,
            };
        }
      },
    ),
  );
}

export function typeQuery(query: TypeQuery) {
  const conditions = [];
  if (query.schemaID) {
    conditions.push(sql`ns.oid = ${query.schemaID}`);
  }
  if (query.schemaName) {
    conditions.push(sql`ns.nspname = ${query.schemaName}`);
  }
  if (query.typeID) {
    conditions.push(sql`ty.oid = ${query.typeID}`);
  }
  if (query.typeName) {
    conditions.push(sql`ty.typname = ${query.typeName}`);
  }
  if (query.category) {
    conditions.push(sql`ty.typcategory = ${query.category}`);
  }
  return conditions;
}
