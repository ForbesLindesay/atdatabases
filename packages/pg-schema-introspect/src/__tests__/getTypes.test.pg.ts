import connect, {sql, DataTypeID} from '@databases/pg';
import getTypes from '../getTypes';
import TypeCateogry from '../enums/TypeCategory';
import {readFileSync, writeFileSync} from 'fs';
import TypeKind from '../enums/TypeKind';
const prettier = require('prettier');

jest.setTimeout(30_000);

const db = connect({bigIntMode: 'number'});

afterAll(async () => {
  await db.dispose();
});

const PG_DATA_TYPE_FILENAME = `${__dirname}/../../../pg-data-type-id/src/index.ts`;
const DEFAULT_TYPESCRIPT_MAPPING_FILENAME = `${__dirname}/../../../pg-schema-print-types/src/DefaultTypeScriptMapping.ts`;

// int8 default for pg-promise is string
// but @databases/pg uses number

const INTERVAL =
  '{years: number,months: number,days: number,hours: number,minutes: number,seconds: number,milliseconds: number, toPostgres: () => string, toISO: () => string}';

const typeMappings: {[key in DataTypeID]?: string} = {
  [DataTypeID.bool]: 'boolean',
  [DataTypeID.bytea]: 'Buffer',
  [DataTypeID.circle]: '{x: number, y: number, radius: number}',
  [DataTypeID.float4]: 'number',
  [DataTypeID.int2]: 'number',
  [DataTypeID.int8]: 'number',
  [DataTypeID.interval]: INTERVAL,
  [DataTypeID.json]: 'any',
  [DataTypeID.jsonb]: 'any',
  [DataTypeID.point]: '{x: number, y: number}',
  [DataTypeID.timestamp]: 'Date',
  [DataTypeID._bool]: 'Array<boolean | nul>',
  [DataTypeID._bytea]: 'Array<Buffer | null>',
  [DataTypeID._int2]: 'Array<number | null>',
  [DataTypeID._int8]: 'Array<number | null>',
  [DataTypeID._interval]: `Array<${INTERVAL} | null>`,
  [DataTypeID._json]: 'Array<any | null>',
  [DataTypeID._numeric]: 'Array<number | null>',
  [DataTypeID._point]: 'Array<{x: number, y: number} | null>',
  [DataTypeID._text]: 'Array<string | null>',
  [DataTypeID._timestamp]: 'Array<Date | null>',
};

async function writeIfDifferent(filename: string, content: string) {
  const prettierOptions = (await prettier.resolveConfig(filename)) || {};
  prettierOptions.parser = 'typescript';
  const formatted = prettier.format(content, prettierOptions);
  let currentContent = '';
  try {
    currentContent = readFileSync(filename, 'utf8');
    if (currentContent === formatted) {
      return;
    }
  } catch (ex: any) {
    if (ex.code !== 'ENOENT') throw ex;
  }
  if (process.env.CI) {
    expect(formatted).toEqual(currentContent);
  }
  writeFileSync(filename, formatted);
}

interface BuiltinTypesState {
  types: {
    subtypeName?: string | undefined;
    pgVersion: [number, number];
    kind: TypeKind;
    typeID: number;
    typeName: string;
    category: TypeCateogry;
    comment: string | null;
  }[];
  ambiguousTypes: {
    [name: string]: {
      subtypeName?: string | undefined;
      pgVersion: [number, number];
      kind: TypeKind;
      typeID: number;
      typeName: string;
      category: TypeCateogry;
      comment: string | null;
    }[];
  };
}
async function writeJsonIfDifferent(
  filename: string,
  content: BuiltinTypesState,
) {
  const formatted = JSON.stringify(content, null, '  ');
  let currentContent = '';
  try {
    currentContent = readFileSync(filename, 'utf8');
    if (currentContent === formatted) {
      return;
    }
  } catch (ex: any) {
    if (ex.code !== 'ENOENT') throw ex;
  }
  if (process.env.CI) {
    expect(formatted).toEqual(currentContent);
  }
  writeFileSync(filename, formatted);
}

test('get built in types', async () => {
  const pgVersion = await getPgVersion();
  const builtInTypesFromPg = (await getTypes(db, {schemaName: 'pg_catalog'}))
    .filter((t) => t.kind !== TypeKind.Composite)
    .map((t) => ({
      pgVersion,
      kind: t.kind,
      typeID: t.typeID,
      typeName: t.typeName,
      category: t.category,
      comment: t.comment,
      ...('subtypeName' in t ? {subtypeName: t.subtypeName} : {}),
    }));

  const oldState: BuiltinTypesState = JSON.parse(
    readFileSync(`${__dirname}/builtinTypes.json`, 'utf8'),
  );
  let {types: builtInTypesFromFile} = oldState;
  const {ambiguousTypes} = oldState;
  for (const typeFromPg of builtInTypesFromPg) {
    const ambiguousType = ambiguousTypes[typeFromPg.typeName] || [];
    if (process.env.CI) {
      if (!ambiguousType.length) {
        const typeFromFile = builtInTypesFromFile.find(
          (typeFromFile) => typeFromFile.typeID === typeFromPg.typeID,
        );
        if (typeFromFile) {
          if (typeFromPg.pgVersion[0] >= typeFromFile.pgVersion[0]) {
            expect(typeFromFile).toEqual(typeFromPg);
          }
        } else {
          expect(builtInTypesFromFile).toContainEqual(typeFromPg);
        }
      }
    } else if (ambiguousType.length) {
      let found = false;
      const existingTypes = ambiguousType.filter((t) => {
        if (t.typeID === typeFromPg.typeID) {
          if (lte(t.pgVersion, typeFromPg.pgVersion)) {
            return false;
          } else {
            found = true;
          }
        }
        return true;
      });
      if (!found) {
        ambiguousTypes[typeFromPg.typeName] = sortByPostgresVersion([
          ...existingTypes,
          typeFromPg,
        ]);
      }
    } else {
      // if there are missing types, you can add them by running
      // with PG_TEST_IMAGE=postgres:10.14-alpine (replacing with the relevant version)
      let found = false;
      builtInTypesFromFile = builtInTypesFromFile.filter((typeFromFile) => {
        if (typeFromFile.typeName === typeFromPg.typeName) {
          if (typeFromFile.typeID !== typeFromPg.typeID) {
            found = true;
            ambiguousTypes[typeFromFile.typeName] = sortByPostgresVersion([
              typeFromFile,
              typeFromPg,
            ]);
            return false;
          } else if (lte(typeFromFile.pgVersion, typeFromPg.pgVersion)) {
            return false;
          } else {
            found = true;
          }
        }
        return true;
      });
      if (!found) {
        builtInTypesFromFile.push(typeFromPg);
      }
    }
  }

  builtInTypesFromFile.sort((a, b) => (a.typeName > b.typeName ? 1 : -1));

  await writeJsonIfDifferent(`${__dirname}/builtinTypes.json`, {
    ambiguousTypes,
    types: builtInTypesFromFile,
  });

  const groupedTypes = builtInTypesFromFile.reduce<{
    [key: string]: (typeof builtInTypesFromFile)[number][];
  }>((result, ty) => {
    const category = Object.keys(TypeCateogry).find(
      (c) => (TypeCateogry as any)[c] === ty.category,
    )!;
    result[category] = (result[category] || []).concat([ty]);
    return result;
  }, {});
  expect(
    Object.keys(groupedTypes)
      .sort()
      .reduce<{[key: string]: string[]}>((result, key) => {
        return {
          ...result,
          [key]: groupedTypes[key].map(
            (ty) =>
              `${ty.typeID} = ${ty.typeName}` +
              ('subtypeName' in ty && ty.subtypeName
                ? `<${ty.subtypeName}>`
                : ''),
          ),
        };
      }, {}),
  ).toMatchSnapshot();

  const PgDataTypeIDsEnum = [
    '// auto generated by test suite of pg-schema-introspect',
    '',
    'enum PgDataTypeID {',
  ];
  Object.keys(groupedTypes).forEach((groupName, i) => {
    if (i !== 0) PgDataTypeIDsEnum.push('');
    PgDataTypeIDsEnum.push(`  // === ${groupName} ===`);
    groupedTypes[groupName].forEach((type) => {
      PgDataTypeIDsEnum.push('');
      const commentLines = [];
      if (type.comment) {
        commentLines.push(type.comment);
      }
      if (type.kind === TypeKind.Array) {
        commentLines.push(`Array<${type.subtypeName}>`);
      }
      if (commentLines.length) {
        PgDataTypeIDsEnum.push(`  /**`);
        commentLines.forEach((commentLine, j) => {
          if (j !== 0) {
            PgDataTypeIDsEnum.push(`   *`);
          }
          PgDataTypeIDsEnum.push(`   * ${commentLine}`);
        });
        PgDataTypeIDsEnum.push(`   */`);
      }
      PgDataTypeIDsEnum.push(`  ${type.typeName} = ${type.typeID},`);
    });
  });
  PgDataTypeIDsEnum.push(`}`);
  PgDataTypeIDsEnum.push(``);
  PgDataTypeIDsEnum.push(`export default PgDataTypeID;`);
  PgDataTypeIDsEnum.push(`module.exports = PgDataTypeID;`);
  PgDataTypeIDsEnum.push(`module.exports.default = PgDataTypeID;`);
  PgDataTypeIDsEnum.push(``);
  await writeIfDifferent(PG_DATA_TYPE_FILENAME, PgDataTypeIDsEnum.join('\n'));

  const pgTypes = require('pg-types/lib/textParsers');
  const mapping = new Map<number, unknown>();
  const reverseMapping = new Map<unknown, number[]>();
  pgTypes.init((id: number, parser: unknown) => {
    mapping.set(id, parser);
    const m = reverseMapping.get(parser) || [];
    reverseMapping.set(parser, [...m, id]);
  });
  const typeMappingLines: string[] = [];
  mapping.forEach((parser, id) => {
    const allIDs = reverseMapping.get(parser) || [];
    const idsWithMapping = allIDs.filter((typeID) => typeID in typeMappings);
    if (idsWithMapping.length === 0) {
      throw new Error(
        'There is no mapping for: ' +
          allIDs.map((typeID) => DataTypeID[typeID]).join(', '),
      );
    }
    if (idsWithMapping.length > 1) {
      throw new Error(
        'There is ambiguity between: ' +
          idsWithMapping.map((typeID) => DataTypeID[typeID]).join(', '),
      );
    }
    typeMappingLines.push(
      `  [DataTypeID.${DataTypeID[id]}, '${
        (typeMappings as any)[idsWithMapping[0]]
      }'],`,
    );
  });
  const DefaultTypeScriptMapping = [
    '// auto generated by test suite of pg-schema-introspect',
    ``,
    `import DataTypeID from '@databases/pg-data-type-id';`,
    ``,
    `const DefaultTypeScriptMapping = new Map([`,
    ...typeMappingLines.sort(),
    `]);`,
    ``,
    `export default DefaultTypeScriptMapping;`,
  ];
  await writeIfDifferent(
    DEFAULT_TYPESCRIPT_MAPPING_FILENAME,
    DefaultTypeScriptMapping.join('\n'),
  );
});

test('get custom types', async () => {
  await db.query(sql`CREATE SCHEMA gettypes`);
  await db.query(
    sql`
      CREATE TYPE gettypes.currency AS ENUM('USD', 'GBP');
      COMMENT ON TYPE gettypes.currency IS 'Three character currency code';

      CREATE DOMAIN gettypes.email AS TEXT CHECK (VALUE ~ '^.+@.+$');
      COMMENT ON TYPE gettypes.email IS 'An email address';

      CREATE TYPE gettypes.money_with_currency AS (
        value MONEY,
        currency gettypes.currency
      );
      COMMENT ON TYPE gettypes.money_with_currency IS 'A monetary value with currency';
    `,
  );
  await db.query(
    sql`
      CREATE TABLE gettypes.tab (
        email gettypes.email NOT NULL PRIMARY KEY,
        money gettypes.money_with_currency
      );
    `,
  );
  await db.query(
    sql`
      INSERT INTO gettypes.tab (email, money) VALUES (${'forbes@lindesay.co.uk'}, ROW (${10}, 'USD'))
    `,
  );
  expect(await db.query(sql`SELECT * FROM gettypes.tab`))
    .toMatchInlineSnapshot(`
    Array [
      Object {
        "email": "forbes@lindesay.co.uk",
        "money": "($10.00,USD)",
      },
    ]
  `);
  expect(
    (await getTypes(db, {schemaName: 'gettypes'}))
      .map((t) => {
        const result = {
          ...t,
          schemaID: typeof t.schemaID === 'number' ? '<oid>' : t.schemaID,
          typeID: typeof t.typeID === 'number' ? '<oid>' : t.typeID,
        };
        if ('subtypeID' in result && typeof result.subtypeID === 'number') {
          result.subtypeID = '<oid>' as any;
        }
        if ('basetypeID' in result && typeof result.basetypeID === 'number') {
          result.basetypeID = '<oid>' as any;
        }
        if ('classID' in result && typeof result.classID === 'number') {
          result.classID = '<oid>' as any;
        }
        if ('attributes' in result) {
          result.attributes = result.attributes.map((a) => ({
            ...a,
            classID: typeof a.classID === 'number' ? '<oid>' : a.classID,
            schemaID: typeof a.schemaID === 'number' ? '<oid>' : a.schemaID,
            typeID: typeof a.typeID === 'number' ? '<oid>' : a.typeID,
          })) as any[];
        }
        return result;
      })
      .filter((t) => {
        // newer versions of postgres include this, but we need the tests to continue
        // passing on postgres 10.14 for now.
        return t.typeName !== '_email';
      }),
  ).toMatchInlineSnapshot(`
    Array [
      Object {
        "category": "A",
        "comment": null,
        "kind": "array",
        "schemaID": "<oid>",
        "schemaName": "gettypes",
        "subtypeID": "<oid>",
        "subtypeName": "currency",
        "typeID": "<oid>",
        "typeName": "_currency",
      },
      Object {
        "category": "A",
        "comment": null,
        "kind": "array",
        "schemaID": "<oid>",
        "schemaName": "gettypes",
        "subtypeID": "<oid>",
        "subtypeName": "money_with_currency",
        "typeID": "<oid>",
        "typeName": "_money_with_currency",
      },
      Object {
        "category": "A",
        "comment": null,
        "kind": "array",
        "schemaID": "<oid>",
        "schemaName": "gettypes",
        "subtypeID": "<oid>",
        "subtypeName": "tab",
        "typeID": "<oid>",
        "typeName": "_tab",
      },
      Object {
        "category": "E",
        "comment": "Three character currency code",
        "kind": "e",
        "schemaID": "<oid>",
        "schemaName": "gettypes",
        "typeID": "<oid>",
        "typeName": "currency",
        "values": Array [
          "GBP",
          "USD",
        ],
      },
      Object {
        "basetypeID": "<oid>",
        "basetypeName": "text",
        "category": "S",
        "comment": "An email address",
        "kind": "d",
        "schemaID": "<oid>",
        "schemaName": "gettypes",
        "typeID": "<oid>",
        "typeName": "email",
      },
      Object {
        "attributes": Array [
          Object {
            "attributeName": "currency",
            "attributeNumber": 2,
            "classID": "<oid>",
            "className": "money_with_currency",
            "comment": null,
            "default": null,
            "hasDefault": false,
            "notNull": false,
            "schemaID": "<oid>",
            "schemaName": "gettypes",
            "typeID": "<oid>",
            "typeLength": -1,
          },
          Object {
            "attributeName": "value",
            "attributeNumber": 1,
            "classID": "<oid>",
            "className": "money_with_currency",
            "comment": null,
            "default": null,
            "hasDefault": false,
            "notNull": false,
            "schemaID": "<oid>",
            "schemaName": "gettypes",
            "typeID": "<oid>",
            "typeLength": -1,
          },
        ],
        "category": "C",
        "classID": "<oid>",
        "comment": "A monetary value with currency",
        "kind": "c",
        "schemaID": "<oid>",
        "schemaName": "gettypes",
        "typeID": "<oid>",
        "typeName": "money_with_currency",
      },
      Object {
        "attributes": Array [
          Object {
            "attributeName": "email",
            "attributeNumber": 1,
            "classID": "<oid>",
            "className": "tab",
            "comment": null,
            "default": null,
            "hasDefault": false,
            "notNull": true,
            "schemaID": "<oid>",
            "schemaName": "gettypes",
            "typeID": "<oid>",
            "typeLength": -1,
          },
          Object {
            "attributeName": "money",
            "attributeNumber": 2,
            "classID": "<oid>",
            "className": "tab",
            "comment": null,
            "default": null,
            "hasDefault": false,
            "notNull": false,
            "schemaID": "<oid>",
            "schemaName": "gettypes",
            "typeID": "<oid>",
            "typeLength": -1,
          },
        ],
        "category": "C",
        "classID": "<oid>",
        "comment": null,
        "kind": "c",
        "schemaID": "<oid>",
        "schemaName": "gettypes",
        "typeID": "<oid>",
        "typeName": "tab",
      },
    ]
  `);
});

async function getPgVersion(): Promise<[number, number]> {
  // e.g. PostgreSQL 10.1 on x86_64-apple-darwin16.7.0, compiled by Apple LLVM version 9.0.0 (clang-900.0.38), 64-bit
  const [{version: sqlVersionString}] = await db.query(
    db.sql`SELECT version();`,
  );
  const match = /PostgreSQL (\d+).(\d+)/.exec(sqlVersionString);
  if (match) {
    const [, major, minor] = match;
    return [parseInt(major, 10), parseInt(minor, 10)];
  }
  return [0, 0];
}

function lte(a: [number, number], b: [number, number]) {
  return a[0] < b[0] || (a[0] === b[0] && (a[1] === b[1] || a[1] < b[1]));
}

function sortByPostgresVersion<T extends {pgVersion: [number, number]}>(
  records: readonly T[],
) {
  return records
    .slice()
    .sort(
      (a, b) =>
        a.pgVersion[0] - b.pgVersion[0] || a.pgVersion[1] - b.pgVersion[1],
    );
}
