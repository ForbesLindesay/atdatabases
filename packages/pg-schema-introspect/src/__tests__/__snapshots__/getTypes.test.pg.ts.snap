// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get built in types 1`] = `
Object {
  "Array": Array [
    "1023 = _abstime<abstime>",
    "1034 = _aclitem<aclitem>",
    "1561 = _bit<bit>",
    "1000 = _bool<bool>",
    "1020 = _box<box>",
    "1014 = _bpchar<bpchar>",
    "1001 = _bytea<bytea>",
    "1002 = _char<char>",
    "1012 = _cid<cid>",
    "651 = _cidr<cidr>",
    "719 = _circle<circle>",
    "1263 = _cstring<cstring>",
    "1182 = _date<date>",
    "3913 = _daterange<daterange>",
    "1021 = _float4<float4>",
    "1022 = _float8<float8>",
    "3644 = _gtsvector<gtsvector>",
    "1041 = _inet<inet>",
    "1005 = _int2<int2>",
    "1006 = _int2vector<int2vector>",
    "1007 = _int4<int4>",
    "3905 = _int4range<int4range>",
    "1016 = _int8<int8>",
    "3927 = _int8range<int8range>",
    "1187 = _interval<interval>",
    "199 = _json<json>",
    "3807 = _jsonb<jsonb>",
    "4073 = _jsonpath<jsonpath>",
    "629 = _line<line>",
    "1018 = _lseg<lseg>",
    "1040 = _macaddr<macaddr>",
    "775 = _macaddr8<macaddr8>",
    "791 = _money<money>",
    "1003 = _name<name>",
    "1231 = _numeric<numeric>",
    "3907 = _numrange<numrange>",
    "1028 = _oid<oid>",
    "1013 = _oidvector<oidvector>",
    "1019 = _path<path>",
    "3221 = _pg_lsn<pg_lsn>",
    "5039 = _pg_snapshot<pg_snapshot>",
    "1017 = _point<point>",
    "1027 = _polygon<polygon>",
    "2201 = _refcursor<refcursor>",
    "2210 = _regclass<regclass>",
    "4192 = _regcollation<regcollation>",
    "3735 = _regconfig<regconfig>",
    "3770 = _regdictionary<regdictionary>",
    "4090 = _regnamespace<regnamespace>",
    "2208 = _regoper<regoper>",
    "2209 = _regoperator<regoperator>",
    "1008 = _regproc<regproc>",
    "2207 = _regprocedure<regprocedure>",
    "4097 = _regrole<regrole>",
    "2211 = _regtype<regtype>",
    "1024 = _reltime<reltime>",
    "1009 = _text<text>",
    "1010 = _tid<tid>",
    "1183 = _time<time>",
    "1115 = _timestamp<timestamp>",
    "1185 = _timestamptz<timestamptz>",
    "1270 = _timetz<timetz>",
    "1025 = _tinterval<tinterval>",
    "3645 = _tsquery<tsquery>",
    "3909 = _tsrange<tsrange>",
    "3911 = _tstzrange<tstzrange>",
    "3643 = _tsvector<tsvector>",
    "2949 = _txid_snapshot<txid_snapshot>",
    "2951 = _uuid<uuid>",
    "1563 = _varbit<varbit>",
    "1015 = _varchar<varchar>",
    "1011 = _xid<xid>",
    "271 = _xid8<xid8>",
    "143 = _xml<xml>",
    "22 = int2vector<int2>",
    "30 = oidvector<oid>",
  ],
  "BitString": Array [
    "1560 = bit",
    "1562 = varbit",
  ],
  "Boolean": Array [
    "16 = bool",
  ],
  "DateTime": Array [
    "702 = abstime",
    "1082 = date",
    "1083 = time",
    "1114 = timestamp",
    "1184 = timestamptz",
    "1266 = timetz",
  ],
  "Geometric": Array [
    "603 = box<point>",
    "718 = circle",
    "628 = line<float8>",
    "601 = lseg<point>",
    "602 = path",
    "600 = point<float8>",
    "604 = polygon",
  ],
  "NetworkAddress": Array [
    "650 = cidr",
    "869 = inet",
  ],
  "Numeric": Array [
    "700 = float4",
    "701 = float8",
    "21 = int2",
    "23 = int4",
    "20 = int8",
    "790 = money",
    "1700 = numeric",
    "26 = oid",
    "2205 = regclass",
    "4191 = regcollation",
    "3734 = regconfig",
    "3769 = regdictionary",
    "4089 = regnamespace",
    "2203 = regoper",
    "2204 = regoperator",
    "24 = regproc",
    "2202 = regprocedure",
    "4096 = regrole",
    "2206 = regtype",
  ],
  "PseudoTypes": Array [
    "2287 = _record",
    "2276 = any",
    "2277 = anyarray",
    "5077 = anycompatible",
    "5078 = anycompatiblearray",
    "5079 = anycompatiblenonarray",
    "5080 = anycompatiblerange",
    "2283 = anyelement",
    "3500 = anyenum",
    "2776 = anynonarray",
    "3831 = anyrange",
    "2275 = cstring",
    "3838 = event_trigger",
    "3115 = fdw_handler",
    "325 = index_am_handler",
    "2281 = internal",
    "2280 = language_handler",
    "2282 = opaque",
    "32 = pg_ddl_command",
    "2249 = record",
    "269 = table_am_handler",
    "2279 = trigger",
    "3310 = tsm_handler",
    "2278 = void",
  ],
  "Range": Array [
    "3912 = daterange",
    "3904 = int4range",
    "3926 = int8range",
    "3906 = numrange",
    "3908 = tsrange",
    "3910 = tstzrange",
  ],
  "String": Array [
    "1042 = bpchar",
    "18 = char",
    "19 = name<char>",
    "3402 = pg_dependencies",
    "5017 = pg_mcv_list",
    "3361 = pg_ndistinct",
    "194 = pg_node_tree",
    "25 = text",
    "1043 = varchar",
  ],
  "Timespan": Array [
    "1186 = interval",
    "703 = reltime",
    "704 = tinterval",
  ],
  "Unknown": Array [
    "705 = unknown",
  ],
  "UserDefined": Array [
    "1033 = aclitem",
    "17 = bytea",
    "29 = cid",
    "3642 = gtsvector",
    "114 = json",
    "3802 = jsonb",
    "4072 = jsonpath",
    "829 = macaddr",
    "774 = macaddr8",
    "3220 = pg_lsn",
    "5038 = pg_snapshot",
    "1790 = refcursor",
    "210 = smgr",
    "27 = tid",
    "3615 = tsquery",
    "3614 = tsvector",
    "2970 = txid_snapshot",
    "2950 = uuid",
    "28 = xid",
    "5069 = xid8",
    "142 = xml",
  ],
}
`;
